File: code/go/02-structured-output/prompt_template_test.go
package main

import (
	"strings"
	"testing"
)

// ---------------------------------------------------------------------------
// buildSystemPrompt
// ---------------------------------------------------------------------------

func TestBuildSystemPromptContainsFocusArea(t *testing.T) {
	result, err := buildSystemPrompt("practical implementation details")
	if err != nil {
		t.Fatal(err)
	}
	if !strings.Contains(result, "practical implementation details") {
		t.Error("system prompt does not contain the focus area")
	}
}

func TestBuildSystemPromptNoLeftoverBraces(t *testing.T) {
	result, err := buildSystemPrompt("testing")
	if err != nil {
		t.Fatal(err)
	}
	if strings.Contains(result, "{") || strings.Contains(result, "}") {
		t.Errorf("leftover template braces found in: %q", result)
	}
}

func TestBuildSystemPromptEmptyReturnsError(t *testing.T) {
	_, err := buildSystemPrompt("")
	if err == nil {
		t.Error("expected error for empty focusArea, got nil")
	}
}

func TestBuildSystemPromptWhitespaceOnlyReturnsError(t *testing.T) {
	_, err := buildSystemPrompt("   ")
	if err == nil {
		t.Error("expected error for whitespace-only focusArea, got nil")
	}
}

// ---------------------------------------------------------------------------
// buildUserPrompt
// ---------------------------------------------------------------------------

func TestBuildUserPromptContainsArticleText(t *testing.T) {
	result, err := buildUserPrompt("AI agents are fascinating.")
	if err != nil {
		t.Fatal(err)
	}
	if !strings.Contains(result, "AI agents are fascinating.") {
		t.Error("user prompt does not contain the article text")
	}
}

func TestBuildUserPromptNoLeftoverBraces(t *testing.T) {
	result, err := buildUserPrompt("Some article text")
	if err != nil {
		t.Fatal(err)
	}
	if strings.Contains(result, "{") || strings.Contains(result, "}") {
		t.Errorf("leftover template braces found in: %q", result)
	}
}

func TestBuildUserPromptEmptyReturnsError(t *testing.T) {
	_, err := buildUserPrompt("")
	if err == nil {
		t.Error("expected error for empty articleText, got nil")
	}
}

func TestBuildUserPromptWhitespaceOnlyReturnsError(t *testing.T) {
	_, err := buildUserPrompt("   ")
	if err == nil {
		t.Error("expected error for whitespace-only articleText, got nil")
	}
}

func TestBuildUserPromptVeryLongInputDoesNotCrash(t *testing.T) {
	longText := strings.Repeat("word ", 2_000)
	result, err := buildUserPrompt(longText)
	if err != nil {
		t.Fatal(err)
	}
	if !strings.Contains(result, longText) {
		t.Error("long text was not included in the user prompt")
	}
}

// ---------------------------------------------------------------------------
// buildMessages
// ---------------------------------------------------------------------------

func TestBuildMessagesReturnsTwoMessages(t *testing.T) {
	msgs, err := buildMessages("focus", "text")
	if err != nil {
		t.Fatal(err)
	}
	if len(msgs) != 2 {
		t.Errorf("expected 2 messages, got %d", len(msgs))
	}
}

func TestBuildMessagesFirstRoleIsSystem(t *testing.T) {
	msgs, err := buildMessages("focus", "text")
	if err != nil {
		t.Fatal(err)
	}
	if msgs[0].Role != "system" {
		t.Errorf("expected role=system, got %q", msgs[0].Role)
	}
}

func TestBuildMessagesSecondRoleIsUser(t *testing.T) {
	msgs, err := buildMessages("focus", "text")
	if err != nil {
		t.Fatal(err)
	}
	if msgs[1].Role != "user" {
		t.Errorf("expected role=user, got %q", msgs[1].Role)
	}
}

func TestBuildMessagesSystemContainsFocusArea(t *testing.T) {
	msgs, err := buildMessages("unique-focus-area", "text")
	if err != nil {
		t.Fatal(err)
	}
	if !strings.Contains(msgs[0].Content, "unique-focus-area") {
		t.Error("system message does not contain the focus area")
	}
}

func TestBuildMessagesUserContainsArticleText(t *testing.T) {
	msgs, err := buildMessages("focus", "unique-article-content")
	if err != nil {
		t.Fatal(err)
	}
	if !strings.Contains(msgs[1].Content, "unique-article-content") {
		t.Error("user message does not contain the article text")
	}
}

func TestBuildMessagesNoLeftoverBraces(t *testing.T) {
	msgs, err := buildMessages("focus", "text")
	if err != nil {
		t.Fatal(err)
	}
	for _, msg := range msgs {
		if strings.Contains(msg.Content, "{") || strings.Contains(msg.Content, "}") {
			t.Errorf("leftover braces in message: %q", msg.Content)
		}
	}
}

func TestBuildMessagesEmptyFocusReturnsError(t *testing.T) {
	_, err := buildMessages("", "text")
	if err == nil {
		t.Error("expected error for empty focusArea, got nil")
	}
}

func TestBuildMessagesEmptyArticleReturnsError(t *testing.T) {
	_, err := buildMessages("focus", "")
	if err == nil {
		t.Error("expected error for empty articleText, got nil")
	}
}

// ---------------------------------------------------------------------------
// countTokens
// ---------------------------------------------------------------------------

func TestCountTokensReturnsPositiveInteger(t *testing.T) {
	msgs, err := buildMessages("focus area", "article text here")
	if err != nil {
		t.Fatal(err)
	}
	n, err := countTokens(msgs, model)
	if err != nil {
		t.Fatal(err)
	}
	if n <= 0 {
		t.Errorf("expected positive token count, got %d", n)
	}
}

func TestCountTokensLongerInputYieldsMoreTokens(t *testing.T) {
	short, _ := buildMessages("focus", "short text")
	long, _ := buildMessages("focus", "short text "+strings.Repeat("extra content ", 100))
	nShort, _ := countTokens(short, model)
	nLong, _ := countTokens(long, model)
	if nLong <= nShort {
		t.Errorf("longer input should yield more tokens: short=%d long=%d", nShort, nLong)
	}
}

func TestCountTokensVeryLongInputDoesNotCrash(t *testing.T) {
	msgs, _ := buildMessages("focus", strings.Repeat("word ", 2_000))
	n, err := countTokens(msgs, model)
	if err != nil {
		t.Fatal(err)
	}
	if n <= 0 {
		t.Error("expected positive token count for long input")
	}
}

func TestCountTokensEmptyMessagesReturnsReplyPrimerOverhead(t *testing.T) {
	n, err := countTokens([]message{}, model)
	if err != nil {
		t.Fatal(err)
	}
	if n != 3 {
		t.Errorf("expected 3 (reply primer overhead), got %d", n)
	}
}

File: code/nodejs/02-structured-output/prompt_template.test.js
import { describe, it } from "node:test";
import assert from "node:assert/strict";
import {
  buildSystemPrompt,
  buildUserPrompt,
  buildMessages,
  countTokens,
} from "./prompt_template.js";

// ---------------------------------------------------------------------------
// buildSystemPrompt
// ---------------------------------------------------------------------------

describe("buildSystemPrompt", () => {
  it("contains the focus area in the output", () => {
    const result = buildSystemPrompt("practical implementation details");
    assert.ok(result.includes("practical implementation details"));
  });

  it("has no leftover template braces", () => {
    const result = buildSystemPrompt("testing");
    assert.ok(!result.includes("{") && !result.includes("}"));
  });

  it("throws on empty string", () => {
    assert.throws(() => buildSystemPrompt(""), /empty/i);
  });

  it("throws on whitespace-only string", () => {
    assert.throws(() => buildSystemPrompt("   "), /empty/i);
  });
});

// ---------------------------------------------------------------------------
// buildUserPrompt
// ---------------------------------------------------------------------------

describe("buildUserPrompt", () => {
  it("contains the article text in the output", () => {
    const result = buildUserPrompt("AI agents are fascinating.");
    assert.ok(result.includes("AI agents are fascinating."));
  });

  it("has no leftover template braces", () => {
    const result = buildUserPrompt("Some article text");
    assert.ok(!result.includes("{") && !result.includes("}"));
  });

  it("throws on empty string", () => {
    assert.throws(() => buildUserPrompt(""), /empty/i);
  });

  it("throws on whitespace-only string", () => {
    assert.throws(() => buildUserPrompt("   "), /empty/i);
  });

  it("handles very long input without crashing", () => {
    const longText = "word ".repeat(2_000);
    const result = buildUserPrompt(longText);
    assert.ok(result.includes(longText));
  });
});

// ---------------------------------------------------------------------------
// buildMessages
// ---------------------------------------------------------------------------

describe("buildMessages", () => {
  it("returns exactly two messages", () => {
    const msgs = buildMessages("focus", "text");
    assert.equal(msgs.length, 2);
  });

  it("first message role is system", () => {
    const msgs = buildMessages("focus", "text");
    assert.equal(msgs[0].role, "system");
  });

  it("second message role is user", () => {
    const msgs = buildMessages("focus", "text");
    assert.equal(msgs[1].role, "user");
  });

  it("system message contains focus area", () => {
    const msgs = buildMessages("unique-focus-area", "text");
    assert.ok(msgs[0].content.includes("unique-focus-area"));
  });

  it("user message contains article text", () => {
    const msgs = buildMessages("focus", "unique-article-content");
    assert.ok(msgs[1].content.includes("unique-article-content"));
  });

  it("no leftover braces in any message", () => {
    const msgs = buildMessages("focus", "text");
    for (const msg of msgs) {
      assert.ok(!msg.content.includes("{") && !msg.content.includes("}"));
    }
  });

  it("throws on empty focusArea", () => {
    assert.throws(() => buildMessages("", "text"), /empty/i);
  });

  it("throws on empty articleText", () => {
    assert.throws(() => buildMessages("focus", ""), /empty/i);
  });
});

// ---------------------------------------------------------------------------
// countTokens
// ---------------------------------------------------------------------------

describe("countTokens", () => {
  it("returns a positive integer for normal input", () => {
    const msgs = buildMessages("focus area", "article text here");
    assert.ok(countTokens(msgs) > 0);
  });

  it("longer input yields more tokens", () => {
    const short = buildMessages("focus", "short text");
    const long = buildMessages("focus", "short text " + "extra content ".repeat(100));
    assert.ok(countTokens(long) > countTokens(short));
  });

  it("very long input does not crash", () => {
    const msgs = buildMessages("focus", "word ".repeat(2_000));
    assert.ok(countTokens(msgs) > 0);
  });

  it("empty messages list returns integer (reply-primer overhead)", () => {
    const result = countTokens([]);
    assert.equal(typeof result, "number");
    assert.equal(result, 3);
  });
});
